### DISCUSSION TITLE
Inference with RadixAttention rather than Self-Attention，but output weirdly

### DISCUSSION THREAD

**Discussion by walker-ai** [sgl-project/sglang#1960] [Q&A]
I'm trying to adjust my own LLM inference code, I have replaced the self-attention with RadixAttention, and some necessary components. But I found that the output is weird:

For example, when I gave "who are you" and "Which city is the capital of China?" as prompts, it went:

![image](https://github.com/user-attachments/assets/ca0a2b33-804e-4e12-b447-cb577943b2e8)
or

![image](https://github.com/user-attachments/assets/aae6ee32-3e16-47a7-9c0b-d322e39d4185)

My model is official weight from [meta-llama2-7b-chat](https://github.com/meta-llama/llama), and my inference code is based on [pytorch-llama](https://github.com/hkproj/pytorch-llama).

It really confused me. I'm not sure if it's a problem with attention calculation. Anyone can help me to figure out what's going on inside here. 🙏

**Comment by walker-ai** [comment]
@merrymercy Would you do me a favor, thank you for your time! 🙏

**Comment by merrymercy** [comment]
You need to setup many things like kv cache correctly. You can compare it with sglang code .

**Reply by walker-ai** [reply]
> You need to setup many things like kv cache correctly. You can compare it with sglang code .

But I was able to get normal output with radixattention, and from the results it just looks like poor quality answers with tons of duplicates. Does this have nothing to do with the attn calculation, but just with the model or sample method and parameter settings?
